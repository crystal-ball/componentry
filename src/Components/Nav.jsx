import React from 'react';
import { NavLink } from 'react-router-dom';

export default function Nav() {
  return (
    <nav>
      <div>
        <NavLink to="/components/alerts" activeClassName="active">
          Alerts
        </NavLink>
      </div>
      <div>
        <NavLink to="/components/buttons" activeClassName="active">
          Buttons
        </NavLink>
      </div>
      <div>
        <NavLink to="/components/cards" activeClassName="active">
          Cards
        </NavLink>
      </div>
      <div>
        <NavLink to="/components/drawers" activeClassName="active">
          Drawers
        </NavLink>
      </div>
      <div>
        <NavLink to="/components/dropdowns" activeClassName="active">
          Dropdowns
        </NavLink>
      </div>
      <div>
        <NavLink to="/components/icons" activeClassName="active">
          Icons
        </NavLink>
      </div>
      <div>
        <NavLink to="/components/list-groups" activeClassName="active">
          List Groups
        </NavLink>
      </div>
      <div>
        <NavLink to="/components/modals" activeClassName="active">
          Modals
        </NavLink>
      </div>
      <div>
        <NavLink to="/components/popovers" activeClassName="active">
          Popovers
        </NavLink>
      </div>
      <div>
        <NavLink to="/components/tooltips" activeClassName="active">
          Tooltips
        </NavLink>
      </div>
    </nav>
  );
}
