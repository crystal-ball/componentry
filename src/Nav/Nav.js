// @flow
import { BaseButton } from '../Button/Button'
import elementFactory, { type ElementProps } from '../component-factories/element'

export type Props = {
  fill?: boolean,
  justify?: boolean,
  pills?: boolean,
  vertical?: boolean,
} & ElementProps

const makeNav = tabNav =>
  elementFactory(
    tabNav ? 'TabNav' : 'Nav',
    ({ fill, justify, pills, vertical, ...props }) => ({
      tag: 'nav',
      role: tabNav ? 'tablist' : undefined,
      className: {
        nav: true,
        'nav-tabs': tabNav,
        'flex-column': vertical,
        'nav-pills': pills,
        'nav-fill': fill,
        'nav-justified': justify,
      },
      ...props,
    }),
  )

const Nav = makeNav(false)
const TabNav = makeNav(true)

const NavItem = elementFactory('NavItem', ({ active, ...rest }) => {
  const { href, onClick } = rest

  return {
    className: {
      'nav-item': true,
      active,
      'nav-link': href || onClick,
    },
    /* eslint-disable no-nested-ternary */
    tag: href || onClick ? (href ? 'a' : BaseButton) : 'li',
    ...rest,
  }
})

export { TabNav }

// $FlowFixMe
Nav.Item = NavItem
export default Nav
