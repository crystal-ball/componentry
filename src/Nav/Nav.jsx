// @flow
import classNames from 'classnames'

import elementFactory from '../component-factories/element-factory'
import itemFactory from '../component-factories/item-factory'
import type { ElementProps } from '../component-factories/element-factory'

export type Props = {
  fill?: boolean,
  justify?: boolean,
  pills?: boolean,
  vertical?: boolean,
} & ElementProps

const makeNav = tabNav => {
  const options = {
    name: tabNav ? 'TabNav' : 'Nav',
    tag: 'nav',
    computedClassName: (
      ctxClassName,
      propsClassName,
      { fill, justify, pills, vertical },
    ) =>
      classNames('nav', ctxClassName, propsClassName, {
        'nav-tabs': tabNav,
        'flex-column': vertical,
        'nav-pills': pills,
        'nav-fill': fill,
        'nav-justified': justify,
      }),
    clean: ['fill', 'justify', 'pills', 'vertical'],
  }

  if (tabNav) options.role = 'tablist'

  return elementFactory(options)
}

const Nav = makeNav(false)
const TabNav = makeNav(true)

const NavItem = itemFactory({
  name: 'NavItem',
  defaultClasses: 'nav-item',
  triggerClass: 'nav-link',
})

export { TabNav }

// $FlowFixMe
Nav.Item = NavItem
export default Nav
