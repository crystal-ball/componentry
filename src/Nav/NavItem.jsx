import classNames from 'classnames'

import Button from '../Button/Button'
import elementFactory from '../component-factories/element-factory'

export default elementFactory({
  name: 'NavItem',
  computedClassName: (ctxClassName, propsClassName, { active, href, to }) =>
    classNames('nav-item', ctxClassName, propsClassName, {
      active,
      'nav-link': href || to,
    }),
  computedTag: props => (props.onClick ? Button : 'a'),
  clean: ['active'],
})
