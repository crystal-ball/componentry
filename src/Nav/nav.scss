/* ~componentry/src/Nav/nav */

// --------------------------------------------------------
// Nav container (<nav> or <ul>)

.nav {
  display: flex;
  flex-wrap: wrap;
  padding-left: 0;
  margin-bottom: 0;
  list-style: none;
}

// --------------------------------------------------------
// Vertical

.nav-vertical {
  flex-direction: column;
}

// --------------------------------------------------------
// Pills

.nav-pills {
  .nav-item {
    @include border-radius($nav-pills-border-radius);
  }

  .nav-item.active {
    color: $nav-pills-link-active-color;
    background-color: $nav-pills-link-active-bg;
  }
}

// --------------------------------------------------------
// Justified

.nav-fill {
  .nav-item {
    flex: 1 1 auto;
    text-align: center;
  }
}

.nav-justified {
  .nav-item {
    flex-basis: 0;
    flex-grow: 1;
    text-align: center;
  }
}

// --------------------------------------------------------
// Nav items

.nav-item {
  -webkit-appearance: none;
  border: none;
  display: block;
  padding: $nav-item-padding-y $nav-item-padding-x;
  background: transparent;
}

.nav-item-action {
  color: $nav-item-action-color;
  text-align: left; // reset button font

  @include hover-focus {
    color: $nav-item-action-hover-color;
    text-decoration: $nav-item-action-hover-decoration;
  }

  // Disabled state lightens text
  &.disabled {
    color: $nav-item-disabled-color;
    text-decoration: none;
  }
}
