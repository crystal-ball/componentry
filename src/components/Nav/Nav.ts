import { useTheme } from '../Theme/Theme'
import { navClasses } from '../../utils/class-names'
import { element } from '../../utils/element-creator'
import { ComponentBaseProps } from '../../utils/types'

export interface NavProps extends ComponentBaseProps<'nav'> {
  fill?: boolean
  justify?: boolean
  pills?: boolean
  vertical?: boolean
}

export interface NavItemProps extends ComponentBaseProps<'a'> {}

export interface Nav {
  (props: NavProps): React.ReactElement
  displayName: 'Nav'
  /**
   * [Nav item component üìù](https://componentry.design/components/nav)
   */
  Item: React.FC<NavItemProps>
}

/**
 * [Nav component üìù](https://componentry.design/components/nav)
 * @experimental
 */
export const Nav = ((props: NavProps) => {
  return element('Nav', {
    as: 'nav',
    componentCx: navClasses('nav', props),
    role: 'navigation',
    ...useTheme<NavProps>('Nav'),
    ...props,
  })
}) as Nav
Nav.displayName = 'Nav'

Nav.Item = function NavItem(props) {
  const { variant = 'primary', ...merged } = {
    ...useTheme<NavItemProps>('NavItem'),
    ...props,
  }

  return element('NavItem', {
    as: 'a',
    componentCx: `nav-item-${variant}`,
    ...merged,
  })
}
Nav.Item.displayName = 'NavItem'
