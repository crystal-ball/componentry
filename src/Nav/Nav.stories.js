/* global module */

import React from 'react'
import { storiesOf } from '@storybook/react'
import { boolean } from '@storybook/addon-knobs'

import Nav from './Nav'

storiesOf('Nav', module).add('<Nav />', () => (
  <div className='w-50'>
    <Nav
      fill={boolean('fill', false)}
      justify={boolean('justify', false)}
      pills={boolean('pills', false)}
      vertical={boolean('vertical', false)}
    >
      <Nav.Item href='#' active>
        Active
      </Nav.Item>
      <Nav.Item href='#'>React</Nav.Item>
      <Nav.Item href='#'>Redux</Nav.Item>
      <Nav.Item href='#' disabled>
        Styles
      </Nav.Item>
    </Nav>
  </div>
))
