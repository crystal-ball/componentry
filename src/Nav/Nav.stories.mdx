import { DocsTitle, Meta } from '@storybook/addon-docs/blocks'
import { Nav } from './Nav'

<Meta title='Components/Nav' component={Nav} />

<DocsTitle title='Navs' experimental />

<div className='w-50'>
  <Nav fill={false} justify={false} pills={false} vertical={false}>
    <Nav.Item href='#' active>
      Active
    </Nav.Item>
    <Nav.Item href='#'>React</Nav.Item>
    <Nav.Item href='#'>Redux</Nav.Item>
    <Nav.Item href='#' disabled>
      Styles
    </Nav.Item>
  </Nav>
</div>
