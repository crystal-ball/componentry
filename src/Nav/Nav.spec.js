import React from 'react'
import { render, screen } from '@testing-library/react'

import { elementTests } from '../test/element-tests'
import { Nav } from './Nav'

describe('<Nav />', () => {
  elementTests(Nav)

  it('should render .nav-vertical when vertical is passed', () => {
    render(
      <Nav data-testid='nav' vertical>
        <Nav.Item href='#'>Link one</Nav.Item>
      </Nav>,
    )

    expect(screen.getByTestId('nav')).toHaveClass('nav-vertical')
  })

  it('should render .nav-pills when pills is passed', () => {
    render(
      <Nav data-testid='nav' pills>
        <Nav.Item href='#'>Link one</Nav.Item>
      </Nav>,
    )

    expect(screen.getByTestId('nav')).toHaveClass('nav-pills')
  })
})

describe('<Nav.Item />', () => {
  elementTests(Nav.Item)

  it('when active is passed, then active class should render', () => {
    render(
      <Nav>
        <Nav.Item href='#' active>
          React
        </Nav.Item>
      </Nav>,
    )

    expect(screen.getByText('React')).toHaveClass('active')
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Nav /> snapshots', () => {
  it('renders correctly', () => {
    render(
      <Nav data-testid='nav'>
        <Nav.Item href='#' active>
          Active
        </Nav.Item>
        <Nav.Item href='#'>React</Nav.Item>
        <Nav.Item href='#'>Redux</Nav.Item>
        <Nav.Item href='#' disabled>
          Styles
        </Nav.Item>
      </Nav>,
    )

    expect(screen.getByTestId('nav')).toMatchSnapshot()
  })
})
