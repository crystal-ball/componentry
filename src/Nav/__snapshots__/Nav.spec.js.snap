// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> snapshots renders correctly 1`] = `
<nav
  class="ðŸ…²-nav"
  role="navigation"
>
  <a
    class="ðŸ…²-nav-item nav-item-primary active"
    href="#"
  >
    Active
  </a>
  <a
    class="ðŸ…²-nav-item nav-item-primary"
    href="#"
  >
    React
  </a>
  <a
    class="ðŸ…²-nav-item nav-item-primary"
    href="#"
  >
    Redux
  </a>
  <a
    class="ðŸ…²-nav-item nav-item-primary disabled"
    disabled=""
    href="#"
  >
    Styles
  </a>
</nav>
`;
